internal/collector/processor/idbatchqueue/id_batch_queue_test.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package idbatchqueue
+
+import (
+	"encoding/binary"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestBatcherNew(t *testing.T) {
+	tests := []struct {
+		name                      string
+		numBatches                uint64
+		newBatchesInitialCapacity uint64
+		batchChannelSize          uint64
+		wantErr                   error
+	}{
+		{"invalid numBatches", 0, 0, 1, ErrInvalidNumBatches},
+		{"invalid batchChannelSize", 1, 0, 0, ErrInvalidBatchChannelSize},
+		{"valid", 1, 0, 1, nil},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := New(tt.numBatches, tt.newBatchesInitialCapacity, tt.batchChannelSize)
+			if err != tt.wantErr {
+				t.Errorf("New() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != nil {
+				got.Stop()
+			}
+		})
+	}
+}
+
+func TestTypicalConfig(t *testing.T) {
+	concurrencyTest(t, 10, 100, uint64(4*runtime.NumCPU()))
```

Comment:
why is the batchChannelSize expected to typically be 4*runtime.NumCPU()? isn't this too low? perhaps the thread dequeueing from channel doesn't get scheduled for 20-50ms and in the meanwhile a lot of IDs arrive then they would have to block.

Reply:
This number came from finding one with reasonable performance from the benchmark. Any goroutine trying to send to the channel will be preempted when the channel is full this should give an opportunity for the one receiving for the channel to run, and it should run until the channel is empty or a batch it switched (contention on the mutex). That said the risk here is if there are too many goroutines running but that will be a general problem not only for this go routine.